lib/internal/bootstrap_node.js

```diff
@@ -283,6 +282,12 @@
       if (!caught)
         caught = process.emit('uncaughtException', er);
 
+      // After this call is made the stack is gone.
+      // TODO(trevnorris): Do we want to run all after callbacks here to let
+      // them know an error has occurred? Since it may be a little confusing
+      // that the after callbacks were never called.
```

Comment:
In that case, I think we should make a full `async_hooks.makeCallback(id, cb, args)` similar to the C++ embedder API, and just leave the `emitBefore`, `setCurrentId` and `emitAfter` dance as an internal API to be used in `nextTick` and similar performance critical functions. For 3rd party embedders I would guess that the overhead of a `makeCallback` function won't be an issue.

Possibly we could do the same for `newUid` and `emitInit` such that the public `emitInit` returns a new uid.